src/zenml/deployers/docker/docker_deployer.py

```diff
@@ -0,0 +1,739 @@
+                deployment to be deployed.
+
+        Returns:
+            The DeploymentOperationalState object representing the
+            operational state of the deployed deployment.
+
+        Raises:
+            DeploymentProvisionError: if the deployment deployment
+                fails.
+        """
+        snapshot = deployment.snapshot
+        assert snapshot, "Pipeline snapshot not found"
+
+        environment = environment or {}
+        secrets = secrets or {}
+        # Currently, there is no safe way to pass secrets to a docker
+        # container, so we simply merge them into the environment variables.
+        environment.update(secrets)
+
+        settings = cast(
+            DockerDeployerSettings,
+            self.get_settings(snapshot),
+        )
+
+        existing_metadata = DockerDeploymentMetadata.from_deployment(
+            deployment
+        )
+
+        entrypoint = ServingEntrypointConfiguration.get_entrypoint_command()
+
+        entrypoint_kwargs = {
+            SNAPSHOT_ID_OPTION: snapshot.id,
+            PORT_OPTION: 8000,
+        }
+        if deployment.auth_key:
+            entrypoint_kwargs[AUTH_KEY_OPTION] = deployment.auth_key
+
+        arguments = ServingEntrypointConfiguration.get_entrypoint_arguments(
+            **entrypoint_kwargs
+        )
+
+        # Add the local stores path as a volume mount
+        stack.check_local_paths()
+        local_stores_path = GlobalConfiguration().local_stores_path
+        volumes = {
+            local_stores_path: {
+                "bind": local_stores_path,
+                "mode": "rw",
+            }
+        }
+        environment[ENV_ZENML_LOCAL_STORES_PATH] = local_stores_path
+
+        # check if a container already exists for the deployment
+        container = self._get_container(deployment)
+
+        if container:
+            # the container exists, check if it is running
+            if container.status == "running":
+                logger.debug(
+                    f"Container for deployment '{deployment.name}' is "
+                    "already running",
+                )
+                container.stop(timeout=timeout)
+
+            # the container is stopped or in an error state, remove it
+            logger.debug(
+                f"Removing previous container for deployment "
+                f"'{deployment.name}'",
+            )
+            container.remove(force=True)
+
+        logger.debug(
+            f"Starting container for deployment '{deployment.name}'..."
+        )
+
+        assert deployment.snapshot, "Pipeline snapshot not found"
+        image = self.get_image(deployment.snapshot)
+
+        try:
+            self.docker_client.images.get(image)
+        except docker_errors.ImageNotFound:
+            logger.debug(
+                f"Pulling container image '{image}' for deployment "
+                f"'{deployment.name}'...",
+            )
+            self.docker_client.images.pull(image)
```

Comment:
Doesn't this happen automatically when running a container that doesn't exist locally?